kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/EngineYarnModeSubmitter.scala

```diff
@@ -0,0 +1,452 @@
+    val fs = stagingDirPath.getFileSystem(hadoopConf)
+    val destPath = new Path(stagingDirPath, srcPath.getName)
+    info(s"Copying $srcPath to $destPath")
+    fs.copyFromLocalFile(srcPath, destPath)
+    fs.setPermission(destPath, new FsPermission(STAGING_DIR_PERMISSION))
+
+    val destFs = FileSystem.get(destPath.toUri, hadoopConf)
+    val destStatus = destFs.getFileStatus(destPath)
+
+    val destResource = Records.newRecord(classOf[LocalResource])
+    destResource.setType(resType)
+    destResource.setVisibility(LocalResourceVisibility.APPLICATION)
+    destResource.setResource(URL.fromPath(destPath))
+    destResource.setTimestamp(destStatus.getModificationTime)
+    destResource.setSize(destStatus.getLen)
+    localResources(destName) = destResource
+  }
+
+  private[kyuubi] def setupLaunchEnv(kyuubiConf: KyuubiConf): mutable.HashMap[String, String] = {
+    info("Setting up the launch environment for engine AM container")
+    val env = new mutable.HashMap[String, String]()
+
+    kyuubiConf.getAll
+      .filter { case (k, _) => k.startsWith(KyuubiConf.KYUUBI_ENGINE_YARN_MODE_ENV_PREFIX) }
+      .map { case (k, v) =>
+        (k.substring(KyuubiConf.KYUUBI_ENGINE_YARN_MODE_ENV_PREFIX.length + 1), v)
+      }
+      .foreach { case (k, v) => KyuubiHadoopUtils.addPathToEnvironment(env, k, v) }
+
+    addClasspathEntry(buildPath(Environment.PWD.$$(), LOCALIZED_CONF_DIR), env)
+    env.put(Environment.HADOOP_CONF_DIR.name(), buildPath(Environment.PWD.$$(), LOCALIZED_CONF_DIR))
+    env
+  }
+
+  private def createApplicationSubmissionContext(
+      newApp: YarnClientApplication,
+      containerContext: ContainerLaunchContext): ApplicationSubmissionContext = {
+
+    val appContext = newApp.getApplicationSubmissionContext
+    appContext.setApplicationName(s"Apache Kyuubi $engineType Engine")
+    appContext.setQueue(kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_QUEUE))
+    appContext.setAMContainerSpec(containerContext)
+
+    val allTags = new util.HashSet[String]
+    kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_TAGS).foreach { tags =>
+      allTags.addAll(tags.asJava)
+    }
+    appContext.setApplicationTags(allTags)
+
+    kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_MAX_APP_ATTEMPTS) match {
+      case Some(v) => appContext.setMaxAppAttempts(v)
+      case None => debug(s"${ENGINE_DEPLOY_YARN_MODE_MAX_APP_ATTEMPTS.key} is not set. " +
+          "Cluster's default value will be used.")
+    }
+
+    val capability = Records.newRecord(classOf[Resource])
+    capability.setMemorySize(kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_MEMORY))
+    capability.setVirtualCores(kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_CORES))
+    debug(s"Created resource capability for AM request: $capability")
+    appContext.setResource(capability)
+
+    appContext
+  }
+
+  private def monitorApplication(appId: ApplicationId): Unit = {
+    if (kyuubiConf.get(KyuubiConf.SESSION_ENGINE_STARTUP_WAIT_COMPLETION)) {
```

Comment:
If we don't wait for the engine to finish, it seems AM is not cleaning the staging dir `cleanupStagingDir`

Reply:
you were right, we should clean staging dir when we unregister the AM.

Reply:
seem we don't need to support SESSION_ENGINE_STARTUP_WAIT_COMPLETION, is it possible to always act `waitCompletion=false`?

Reply:
I don't currently think of any other effects of `SESSION_ENGINE_STARTUP_WAIT_COMPLETION` is false, Sounds ok.